File: 0627/zmq/add_two_client.py
#!/usr/bin/env python
#

import time
import zmq
import sys

# initialie request argument
i1 = 4
i2 = 7
print(sys.argv, len(sys.argv))
if len(sys.argv) > 1:
    i1 = int(sys.argv[1])
if len(sys.argv) > 2:
    i2 = int(sys.argv[2])
request = {'i1': i1, 'i2': i2}

print("Connecting to 5555")
context = zmq.Context()
socket = context.socket(zmq.REQ)
socket.connect("tcp://localhost:5555")

print("Sending: {}".format(request))
socket.send_json(request)

response = socket.recv_json()
print("Received: {}".format(response))
print("Answer is {}".format(response['i3']))

File: 0627/zmq/add_two_client.cpp
#include <zmqpp/zmqpp.hpp>
#include <jsoncpp/json/json.h>
#include <sstream>

int main(int argc, char *argv[]) {
  // initialize the 0MQ context
  zmqpp::context context;

  // generate a publish socket
  zmqpp::socket socket (context, zmqpp::socket_type::request);

  // open the connection
  std::cout << "Connecting to 5555" << std::endl;
  socket.connect("tcp://localhost:5555");

  // create request message
  Json::Value root;
  root["i1"] = 4;
  root["i2"] = 7;

  // compose a message from JSON object
  std::stringstream ss;
  ss << root;
  printf("Sending: %s\n", ss.str().c_str());

  // send a request
  zmqpp::message request = ss.str();
  socket.send(request);

  // receive the response
  zmqpp::message message;
  // decompose the message
  socket.receive(message);
  std::string text;
  message >> text;

  printf("Received: %s\n", text.c_str());
  Json::Reader reader;
  Json::Value response;
  reader.parse(text, response);
  printf("Answer is %d\n", response["i3"].asInt());
}
